File: Algorithms/Medium/238.Product-Of-Array-Except-Self/ProductOfArrayExceptSelf.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def productExceptSelf_v1(self, nums: list[int]) -> list[int]:
        n = len(nums)
        result = [1] * n 
        prefix, postfix = [1] * (n + 1), [1] * (n + 1)

        for i in range(n):
            prefix[i + 1] = prefix[i] * nums[i]

        for i in range(n - 1, -1, -1):
            postfix[i] = postfix[i + 1] * nums[i]

        for i in range(n):
            result[i] = prefix[i] * postfix[i + 1]

        return result
    

    # Time complexity: O(n)
    # Space complexity: O(1)
    def productExceptSelf_v2(self, nums: list[int]) -> list[int]:
        n = len(nums)
        result = [1] * n 
        
        product = 1
        for i in range(n - 1):
            product *= nums[i]
            result[i + 1] = product

        product = 1
        for i in range(n - 1, 0, -1):
            product *= nums[i]
            result[i - 1] *= product

        return result

File: Algorithms/Medium/238.Product-Of-Array-Except-Self/ProductOfArrayExceptSelf.go
package medium

// Time complexity: O(n)
// Space complexity: O(n)
func productExceptSelf(nums []int) []int {
	result := make([]int, len(nums))
	prefix := make([]int, len(nums)+1)
	suffix := make([]int, len(nums)+1)

	for i := range prefix {
		prefix[i] = 1
		suffix[i] = 1
	}

	for i := range nums {
		prefix[i+1] = prefix[i] * nums[i]
	}

	for i := len(nums) - 1; i > 0; i-- {
		suffix[i] = suffix[i+1] * nums[i]
	}

	for i := 0; i < len(nums); i++ {
		result[i] = prefix[i] * suffix[i+1]
	}

	return result
}
